feat(api): add setStoredLabware command

sfoster1 opened the pull request:
Adds the new setStoredLabware command to constrain the flex stacker labware pool.

The idea here is that you configure each stacker with some concept of "this is the kind of labware you contain", and an amount of labware. Then, you update how much it contains and when you retrieve stuff `retrieve` creates a new labware object. Only labware that entered the stacker through `store` retains its individuality. 

This PR creates the command and adds the new data to the stacker substate, but it's not yet used for anything; static mode remains, loading labware objects into the hopper remains. In followup prs, we'll remove that older loading capability, and switch things over to using the new way of doing things.

## testing
- automated tests, really; the command does nothing

## reviews
- structure and arguments and such

Closes EXEC-1212

sfoster1 committed:
feat(api): add setStoredLabware command

Adds the new setStoredLabware command to constrain the flex stacker
labware pool.

Closes EXEC-1212

vegano1 commented on api/src/opentrons/protocol_engine/commands/flex_stacker/store.py:
```diff
@@ -88,19 +88,22 @@ async def execute(self, params: StoreParams) -> SuccessData[StoreResult]:
+                self._state_view.labware.get_definition(lid_id),
+                labware_defs[0],
+            ]
+        stack_height = self._state_view.geometry.get_height_of_labware_stack(
```
nice

vegano1 commented on api/src/opentrons/protocol_engine/execution/equipment.py:
```diff
@@ -127,6 +127,39 @@ def __init__(
             or pipette_data_provider.VirtualPipetteDataProvider()
         )
 
+    async def load_definition_for_details(
```
typo? should be `load_definition_FROM_details `

sfoster1 replied:
I guess. I read it as "load the definition for the labware with these details". happy to change it if you want me to

vegano1 reviewed (approved):
The changes look good to me.

CaseyBatten commented on api/src/opentrons/protocol_engine/commands/flex_stacker/set_stored_labware.py:
```diff
@@ -0,0 +1,184 @@
+        # TODO: propagate the limit on max height of the stacker
+        count = min(params.initialCount, 5)
+
+        state_update = (
+            update_types.StateUpdate()
+            .update_flex_stacker_labware_pool_definition(
+                params.moduleId, labware_def, adapter_def, lid_def
+            )
+            .update_flex_stacker_labware_pool_count(params.moduleId, count)
+        )
```
Nice, makes sense.

CaseyBatten commented on api/src/opentrons/protocol_engine/state/module_substates/flex_stacker_substate.py:
```diff
@@ -33,6 +39,18 @@ def new_from_state_change(
         if update.in_static_mode != NO_CHANGE:
             new_mode = update.in_static_mode
 
+        pool_primary_definition = self.pool_primary_definition
+        pool_adapter_definition = self.pool_adapter_definition
+        pool_lid_definition = self.pool_lid_definition
+        if update.pool_constraint != NO_CHANGE:
+            pool_primary_definition = update.pool_constraint.primary_definition
+            pool_adapter_definition = update.pool_constraint.adapter_definition
+            pool_lid_definition = update.pool_constraint.lid_definition
+
+        pool_count = self.pool_count
+        if update.pool_count != NO_CHANGE:
+            pool_count = update.pool_count
+
```
Looks solid

CaseyBatten reviewed (approved):
Looks great, state update implementation was straightforward.

github-actions[bot] committed:
fix(analyses-snapshot-testing): heal exec-1212-set-stored-labware-pe snapshots (#17552)

This PR was requested on the PR
https://github.com/Opentrons/opentrons/pull/17540

Co-authored-by: sfoster1 <3091648+sfoster1@users.noreply.github.com>
